src/tqec/computation/block_graph.py

```diff
@@ -0,0 +1,257 @@
+from copy import deepcopy
+from io import BytesIO
+
+from tqec.computation.base_graph import ComputationGraph
+from tqec.exceptions import TQECException
+from tqec.position import Direction3D, SignedDirection3D
+from tqec.computation.cube import Cube, CubeKind
+from tqec.computation.pipe import Pipe, PipeKind
+from tqec.computation.zx_graph import ZXGraph
+from tqec.computation.correlation import CorrelationSurface
+
+if TYPE_CHECKING:
+    from tqec.interop.collada_html_viewer import ColladaHTMLViewer
+    from tqec.computation.abstract_observable import AbstractObservable
+    from tqec.interop.color import RGBA
+    from tqec.interop.geometry import FaceKind
+
+
+BlockKind = CubeKind | PipeKind
+"""Valid block types in the library."""
+
+
+class BlockGraph(ComputationGraph[Cube, Pipe]):
+    """An undirected graph representation of a logical computation.
+
+    The block graph specifies the explicit block structures of the spacetime diagram, including
+    the layout of the code patches, the boundary types and the connectivity of the patches
+    across the spacetime. The block graph can be compiled and used to generate the concrete
+    circuits that implement the logical computation.
+
+    The nodes in the graph are the cubes that represent a unit of code patch in the spacetime.
+    The edges in the graph are the pipes that connect the cubes. The pipes guide the logical
+    computation and the information flow between the code patches over spacetime.
+    """
+
+    def add_edge(self, u: Cube, v: Cube, kind: PipeKind | None = None) -> None:
+        """Add an edge to the graph. If the nodes do not exist in the graph, the nodes will be
+        created and added to the graph.
+
+        Args:
+            edge: The edge to add to the graph.
+
+        Raises:
+            TQECException: For each node in the edge, if there is already a node which is not
+                equal to it at the same position, or the node is a port but there is already a
+                different port with the same label in the graph.
+        """
+        if kind is None:
+            pipe = Pipe.from_cubes(u, v)
+        else:
+            pipe = Pipe(u, v, kind)
+        self._add_edge_and_nodes_with_checks(u, v, pipe)
+
+    def validate(self) -> None:
+        """Check the validity of the block structures represented by the graph.
+
+        Refer to the Fig.9 in arXiv:2404.18369. Currently, we ignore the b) and e),
+        only check the following conditions:
+        a). no fanout: ports can only have one pipe connected to them.
+        c). time-like Y: Y cubes can only have time-like pipes connected to them.
+        d). no 3D corner: a cube cannot have pipes in all three directions.
+        f). match color at passthrough: two pipes in a "pass-through" should have the same
+        color orientation.
+        g). match color at turn: two pipes in a "turn" should have the matching colors on
+        faces that are touching.
```

Comment:
Will likely not render correctly in the documentation. Maybe 
```suggestion
        Refer to the Fig.9 in arXiv:2404.18369. Currently, we ignore the b) and e),
        only check the following conditions:
        
        a) no fanout: ports can only have one pipe connected to them.
        c) time-like Y: Y cubes can only have time-like pipes connected to them.
        d) no 3D corner: a cube cannot have pipes in all three directions.
        f) match color at passthrough: two pipes in a "pass-through" should have the same
        color orientation.
        g) match color at turn: two pipes in a "turn" should have the matching colors on
        faces that are touching.
```
will, but I am not sure. It might be better to just check that in #381 once this PR is merged and fix there.